src/EditorFeatures/Core/Implementation/ForegroundNotification/ForegroundNotificationService.cs

```diff
@@ -311,9 +323,34 @@ private void Enqueue_NoLock(LinkedListNode<PendingWork> entry)
                         return;
                     }
 
-                    current = current.Previous;
+                    var previous = current.Previous;
+
+                    // eagarly clean up cancelled entry from the list backward
```

Comment:
Why are we cleaning up the queue during a later enqueue versus cleaning it up eagerly when the cancellation was triggered?

Reply:
are you saying we should register action for each cancellation? that's one way.. but that probably require some thought on concurrent issue since cancellation can happen in any thread. 

here I am cleaning up very naively from BG. work that got cancelled might still left in the queue and cleaned up when running in FG. 

removing 10K+ items from list from FG in 15ms is not an issue if that ever happens.

Reply:
> removing 10K+ items from list from FG in 15ms is not an issue if that ever happens.

This is fixing a symptom.  Why are we getting 10k items in the FG queue.  You mentioned it happens just with a simple console app.  Certainly it should be possible to investigate why that is happening and fixing the root cause.

Reply:
rest of PR is fixing the issue we are getting thousands of UI requests. and I already stated that was happening since tagger was moving everything to UI thread without filtering anything on BG.

there was even more I could improve here since tagger can be disposed while its request is on UI thread which can not be discarded. and this make general cancellation cleanup better for this service.